AnyProcedure enhancements

sviatoslav opened the pull request:
This pull requests adds two classes: `AnyInputProcedure<Input>` and `AnyOutputProcedure<Output>`.

During implementation of this classes I replaced `AnyProcedureBox` with `ValueBox`.

Tests for procedures are the same as they were for `AnyProcedure`. Common code was moved to new `BaseAnyProcedureTests`.

sviatoslav commented:
@danthorpe could you please help me with these failures? I don't understand what went wrong.

danthorpe commented:
Hi @sviatoslav - thanks so much for contributing.

Don't worry about CI - there are some changes on Buildkite which means that 3rd party PRs don't work because you don't have a user, but thankfully, one of the contributors has refreshed it. The only failure now, is the integration tests with CocoaPods, which doesn't work for building 3rd party forks.

I haven't had a chance to look through your changes here in detail yet - but will get to it over the weekend.

In the mean time - can you explain the rationale behind your additions here? Why do we need these two new classes? Thanks!

sviatoslav commented:
As far as I understand `AnyProcedure` is workaround for using protocols with associated types as variable.
In my project I have procedure that conforms only to `OutputProcedure` protocol. At first I added `InputProcedure` conformance and used `Pending<Void>` as `input`.
I didn't like this and added these two classes.
I don't use `AnyInputProcedure` but I believe that it might be useful for others.

sviatoslav commented on Sources/ProcedureKit/Any.swift:
```diff
@@ -6,50 +6,108 @@
+
+enum ProcedureOutputBoxCreator {
+    static func outputBox<P: OutputProcedure>(for procedure: P) -> ValueBox<Pending<ProcedureResult<P.Output>>> {
+        let getter = { return procedure.output }
```
When I implemented this I was thinking about adding `[unowned self]`

danthorpe replied:
Yeah, I'm not sure. This is partly why I don't use closures for type erasure.

sviatoslav replied:
It should not change procedure lifecycle. Group procedure holds strong reference to child procedure anyway. But `unowned` also will work. So that's more about code conventions.

sviatoslav commented:
Is there any progress on reviewing it? I'd like to create pod based on ProcedureKit, but I need  `AnyOuputProcedure` for that. If you are not going to merge this pull request please let me know. In this case I'll add this code to my pod.

danthorpe commented:
Hi @sviatoslav - yeah, will merge your PR - thanks for the contribution, and sorry for the delay.

I will probably refactor this a little bit though, so that it follows the same type-erasure pattern as elsewhere in the framework.

sviatoslav commented:
No problems. Af first I was trying to do the same type-erasure pattern, but that required a lot of duplicating code. That's why I used another one.

Thanks for merging.

And one more question does not related to this pull request? Will you update pod? As I wrote before latest version is 4.0.0.beta.4 and it`s seem's to be very old.

```
pod search --simple ProcedureKit
-> ProcedureKit (4.0.0.beta.4)
   Powerful Operation subclasses in Swift.
   pod 'ProcedureKit', '~> 4.0.0.beta.4'
   - Homepage: https://github.com/ProcedureKit/ProcedureKit
   - Source:   https://github.com/ProcedureKit/ProcedureKit.git
   - Versions: 4.0.0.beta.4 [master repo]
   - Subspecs:
     - ProcedureKit/Standard (4.0.0.beta.4)
     - ProcedureKit/Network (4.0.0.beta.4)
     - ProcedureKit/Location (4.0.0.beta.4)
     - ProcedureKit/Cloud (4.0.0.beta.4)
     - ProcedureKit/Mobile (4.0.0.beta.4)
```

jshier commented:
`pod search` only searches your local spec repo, so you may need to run `pod repo update` to get the latest specs.

sviatoslav commented:
@jshier thanks. Updated. Latest version is 4.0.0.beta.5

jshier commented:
Strange, latest is `4.0.0` on my local machine and cocoapods.org.

sviatoslav commented:
Oh yes. It's 4.0.0. I don't know how I missed this version. Thanks for help, @jshier 

```
-> ProcedureKit (4.0.0)
   Powerful Operation subclasses in Swift.
   pod 'ProcedureKit', '~> 4.0.0'
   - Homepage: https://github.com/ProcedureKit/ProcedureKit
   - Source:   https://github.com/ProcedureKit/ProcedureKit.git
   - Versions: 4.0.0, 4.0.0.beta.5, 4.0.0.beta.4 [master repo]
   - Subspecs:
     - ProcedureKit/Standard (4.0.0)
     - ProcedureKit/Network (4.0.0)
     - ProcedureKit/Location (4.0.0)
     - ProcedureKit/Cloud (4.0.0)
     - ProcedureKit/All (4.0.0)
     - ProcedureKit/Mobile (4.0.0)
```

danthorpe commented:
Yeah, unfortunately, there was an issue raised about the beta tag in CocoaPods, and I guess I made a mistake when I did Beta 6, because CocoaPods ended up interpreting it as 4.0.0 - but it's really 4.0.0-beta-6.

I've got Beta 7 to prepare, but I've just been mulling over how to version it without screwing everyone up. Given that it actually have breaking changes which are to put into the 4.0.0 release.
